sw/device/lib/base/hardened.h

```diff
@@ -32,4 +36,132 @@ typedef enum hardened_bool {
   kHardenedBoolFalse = 0x1d4u,
 } hardened_bool_t;
 
+/**
+ * Launders the 32-bit value `val`.
+ *
+ * `launder32()` returns a value identical to the input, while introducing an
+ * optimization barrier that prevents the compiler from learning new information
+ * about the original value, based on operations on the laundered value. This
+ * does not work the other way around: some information that the compiler has
+ * already learned about the value may make it through; this last caveat is
+ * explained below.
+ *
+ * For example, in the following code, LLVM would fold `CHECK()` away, since a
```

Comment:
Sure. I'm going to include a note that this implementation applies *specicically* to LLVM, though. GCC has previously performed (extremely rude) optimizations by peering into inline assembly, so I think it would be unwise to assume it wouldn't do that in the future.

Reply:
Yeah, definitely makes sense to call out what toolchains this is expected to work reliably with.